src/printer.js

```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
     var children = [];
 
     path.map(
-      function(childPath) {
+      function(childPath, i) {
         var child = childPath.getValue();
+        let lastIdx = children.length - 1;
+        let lastChild = children[lastIdx];
 
         if (
           namedTypes.Literal.check(child) && typeof child.value === "string"
         ) {
           if (/\S/.test(child.value)) {
             const beginBreak = child.value.match(/^\s*\n/);
             const endBreak = child.value.match(/\n\s*$/);
+            const beginSpace = child.value.match(/^\s+/);
+            const endSpace = child.value.match(/\s+$/);
+
+            if (beginBreak) {
+              children.push(hardline);
+            } else if (beginSpace) {
+              let prev = util.getLast(children);
+              if (prev && prev.type === "line") {
+                // TODO ifBreak with else
+                children.push("{' '}", softline);
+              } else {
+                children.push(" ");
+              }
+            }
+
+            children.push(child.value.replace(/^\s+|\s+$/g, ""));
+
+            if (endBreak) {
+              children.push(hardline);
+            } else if (endSpace) {
+              children.push(" ");
+            }
 
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
           } else if (/\n/.test(child.value)) {
```

Comment:
I'm also confused about something here: how does this work? JSX elements are not split up per-line. For example:

```js
<div>
  foo
  bar
  baz
</div>
```

The text contents are represented by this node:

```js
Node {
  type: 'JSXText',
  start: 5,
  end: 24,
  loc:
  SourceLocation {
    start: Position { line: 1, column: 5 },
    end: Position { line: 5, column: 0 } },
  extra: null,
  value: '\n  foo\n  bar\n  baz\n'
}
```

I don't understand when we are able to just test for a newline and emit only a line. Shouldn't we handle newlines more elegantly?

Note that we cannot just output this text. For the printer to accurately measure, we need to convert them into line objects:

```
join(line, node.value.split("\n"))
```

Something like that. I swear we used to do that but I don't see it anywhere in here.